### PR TITLE
Allow for limiting the countries returned for the country field

### PR DESCRIPTION
**Pull request by Blackbaud-TrevorBurch**
#81
#87

Both addressed by this ^^

The invalid iso is now handled via the built in validation per a discussion with @Blackbaud-ToddRoberts

### CODE REVIEW

**Comment by blackbaud-ado** [comment]
Preview build at [https://host.nxt.blackbaud.com/skyux-lookup?_sp.spa=pr-limit-countires](https://host.nxt.blackbaud.com/skyux-lookup?_sp.spa=pr-limit-countires)

File: src/app/public/modules/country-field/fixtures/country-field.component.fixture.html [hunk 1]
```diff
@@ -2,9 +2,11 @@
   <form>
     <sky-country-field
       name="countryField"
+      [attr.required]="isRequired ? true : undefined"
       [disabled]="isDisabled ? true : undefined"
       [defaultCountry]="defaultCountry"
-      [attr.required]="isRequired ? true : undefined"
+      [required]="isRequired"
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/country-field/fixtures/country-field.component.fixture.html:8]: Both `required` and `attr.required`?

File: src/app/public/modules/country-field/country-field.component.ts [hunk 2]
```diff
@@ -319,37 +350,82 @@ export class SkyCountryFieldComponent implements ControlValueAccessor, OnDestroy
...
+      .classList
+      .contains('sky-phone-field-country-search');
+
+    /* istanbul ignore else */
+    if (!this.isInPhoneField) {
+      /**
+       * The library we get the country data from includes extra phone properties.
+       * We want to remove these unless we are in a phone field
+       */
+      this.countries.forEach((country: any) => {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/country-field/country-field.component.ts:385]: Is there a type we can use besides `any`?

File: src/app/public/modules/country-field/country-field.component.ts [hunk 3]
```diff
@@ -319,37 +350,82 @@ export class SkyCountryFieldComponent implements ControlValueAccessor, OnDestroy
       });
   }
 
+  private countriesEqual(a: SkyCountryFieldCountry, b: SkyCountryFieldCountry) {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/country-field/country-field.component.ts:353]: Code style, return type.

File: src/app/public/modules/country-field/country-field.component.ts [hunk 4]
```diff
@@ -168,12 +206,28 @@ export class SkyCountryFieldComponent implements ControlValueAccessor, OnDestroy
...
+  }
+
+  /**
+   * Angular lifecycle hook for when the component is initialized
+   * @internal
+   */
+  public ngOnInit(): void {
+
+    // tslint:disable-next-line: no-null-keyword
+    this.ngControl = this.injector.get<NgControl>(NgControl as Type<NgControl>, null);
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/country-field/country-field.component.ts:229]: @Blackbaud-AlexKingman This feels fishy to me; didn't you find out an "approved" way to do this? Don't we get the  `NgControl` in the validator callback?

File: src/app/public/modules/country-field/country-field.component.ts [hunk 5]
```diff
@@ -70,6 +86,19 @@ export class SkyCountryFieldComponent implements ControlValueAccessor, OnDestroy
...
+  @Input()
+  public set supportedCountries(value: string[]) {
+    this._supportedCountries = value;
+
+    if (value && value.length > 0) {
+      this.setupCountries();
+    }
+  }
+
+  public get supportedCountries(): string[] {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/country-field/country-field.component.ts:98]: Does the `string` represent the ISO, the name of the country, etc.? We should probably be more specific with what type of string is expected, `supportedCountryISOs` for example.

File: src/app/public/modules/country-field/country-field.component.ts (hunk 4 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/country-field/country-field.component.ts:229]: Trevor talked this over with me. Apparently, theres no way with our approved method to do validation as well as our own internal custom validation without grabbing the control through the injector.

File: src/app/public/modules/country-field/country-field.component.ts (hunk 2 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/country-field/country-field.component.ts:385]: Looks like its a loosey-goosey array we get from the library: https://github.com/jackocnr/intl-tel-input/blob/baeead220dd3462841595648fad7f5b14c383bd3/src/js/data.js

I did notice there's a types library here: https://www.npmjs.com/package/@types/intl-tel-input that we could install. Think its worth it to get around this `any`?

Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/country-field/country-field.component.ts:385]: Eh, I think I tried pulling that types library in for the RC branch and had trouble with it, unless you can figure it out. If it's too much trouble, we can ignore it.

Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/country-field/country-field.component.ts:385]: Looks like that types library doesn't even export the object that we'd use here, so its a moot point. Lets just stick with `any` for now if you don't mind?

**Comment by blackbaud-ado** [comment]
Preview build at [https://host.nxt.blackbaud.com/skyux-lookup?_sp.spa=pr-limit-countires](https://host.nxt.blackbaud.com/skyux-lookup?_sp.spa=pr-limit-countires)

**Pull request merged**